README.md

```diff
@@ -0,0 +1,275 @@
+	}{
+		Enabled: true,
+		Port:    8080,
+	}
+
+	parsed, err := proteus.MustParse(&params, proteus.WithSources(
+		cfgflags.New(),
+		cfgenv.New("CFG"),
+	))
+	if err != nil {
+		parsed.ErrUsage(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	if params.Enabled {
+		fmt.Printf("Starting HTTP server on :%d\n", params.Port)
+	}
+}
+```
+
+Only token has to be specified:
+
+```bash
+go run *.go -token my-secret-token
+
+# Output: Starting HTTP server on :8080
+```
+
+The token is marked as a secret, which is important to avoid leaking its value.
+
+// TODO: link with doc for `Dump()`, `Usage()` and `ErrUsage()`.
+
+### XTypes
+
+_XTypes_ are types provided by _proteus_ to handle complex types and to provide
+more additional functionality.
+
+```go
+package main
+
+func main() {
+	params := struct {
+		BindPort   uint16
+		PrivKey    *xtypes.RSAPrivateKey
+		RequestLog *xtypes.OneOf
+	}{
+		RequestLog: &xtypes.OneOf{
+			Choices: []string{"none", "basic", "full"},
+			UpdateFn: func(s string) {
+				fmt.Printf("Log level changed to %s", s)
+			},
+		},
+	}
+
+	parsed, err := proteus.MustParse(&params, proteus.WithSources(
+		cfgflags.New(),
+		cfgenv.New("CFG"),
+	))
+	if err != nil {
+		parsed.ErrUsage(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("Starting HTTP server on :%d with an RSA key of size %d, log level %s\n",
+		params.BindPort,
+		params.PrivKey.Value().Size(),
+		params.RequestLog.Value())
+}
+```
+
+### Auto-Generated Usage (a.k.a --help)
+
+To have usage information include the `WithAutoUsage` option:
+
+```go
+func main() {
+	params := struct {
+		Database    dbParams
+		Environment *xtypes.OneOf `param_desc:"What environment the app is running on"`
+		Port        uint16
+	}{
+		Database: defaultDBParams(),
+		Environment: &xtypes.OneOf{
+			Choices: []string{"dev", "stg", "prd"},
+		},
+	}
+
+	parsed, err := proteus.MustParse(&params,
+		proteus.WithAutoUsage(os.Stderr, "Demo App", func() { os.Exit(0) }),
+		proteus.WithSources(
+			cfgflags.New(),
+			cfgenv.New("CFG"),
+		))
+	if err != nil {
+		parsed.ErrUsage(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("Running in %s mode\n", params.Environment.Value())
+}
+
+type dbParams struct {
+	Server   string `param:",optional"    param_desc:"Name of the database server"`
+	Port     uint16 `param:",optional"    param_desc:"TCP port number of database server"`
+	User     string `                     param_desc:"Username for authentication"`
+	Password string `                     param_desc:"Password for authentication"`
+}
+
+func defaultDBParams() dbParams {
+	return dbParams{
+		Server: "localhost",
+		Port:   5432,
+	}
+}
+```
+
+```bash
+go run main.go --help
+
+# Output:
+# Demo App
+# Syntax:
+# ./main \
+#     <-environment (dev|stg|prd)> \
+#     <-port uint16> \
+#     [-help] \
+#   database \
+#     <-password string> \
+#     <-user string> \
+#     [-port uint16] \
+#     [-server string]
```

Comment:
angle brackets are used for required placeholders for values, normally a placeholder Value is shown that describes the expected value, often the same then the parameter name in uppercase.
I think it also makes sense to print "--help" always as first or last parameter.

If you want to follow the common conventionthe generated usage output should be instead:

```suggestion
#     [-help] \
#     -environment dev|stg|prd \
#     -port <PORT> \
#   database \
#     -password <PASSWORD> \
#     -user <USER> \
#     [-port <PORT>] \
#     [-server <SERVER-NAME>]
```

Reply:
I like everything in this suggestion, except `-user <USER> -port <PORT>` (I'm referring only to the repetition). To me it feels better to show it like this:

```
     [-help] \
     -environment dev|stg|prd \
     -port <uint16> \
   database \
     -password <string> \
     -user <string> \
     [-port <uint16>] \
     [-server <string>]
```

Maybe the type can be omitted bellow, when the documentation for each parameter is shown. What do you think? I changed the code to reflect this snippet, while we come up to a consensus.

Reply:
:+1: makes sense for me as well